src/rowbinary/validation.rs

```diff
@@ -295,28 +303,32 @@ impl<'caller, R: Row> SchemaValidator<R> for Option<InnerDataTypeValidator<'_, '
                     }
                     None => {
                         let (full_name, full_data_type) =
-                            inner.root.get_current_column_name_and_type();
-                        panic!(
+                            inner.root.get_current_column_name_and_type()?;
+
+                        Err(Error::SchemaMismatch(format!(
```

Comment:
I also wonder if replacing this with a constructor function marked `#[cold]` could help with mispredicts.